File: src/main/java/j/math/cat/N.java
package j.math.cat;

import java.util.AbstractSet;
import java.util.Iterator;

/**
 * Natural numbers class.
 */
public class N extends AbstractSet<Integer> {
  @Override
  public Iterator<Integer> iterator() {
    return new Iterator<Integer>() {
      int n = 0;

      public boolean hasNext() {
        return true;
      }

      public Integer next() {
        return n++;
      }

      public void remove() {
        throw new UnsupportedOperationException();
      }
    };
  }

  @Override
  public int size() {
    return Integer.MAX_VALUE;
  }
}

File: scala3/src/main/scala/math/sets/N.scala
package math.sets

import math.Base._

/**
 * Set of natural numbers.
 */
object N extends BigSet[BigInt]("ℕ") with EnumerableSet[BigInt]:

  /**
    * Iterator over natural numbers
    * @return an iterator of natural numbers
    */
  override def iterator: Iterator[BigInt] =
    new Iterator[BigInt]:
      private var n: BigInt = -1
  
      override def hasNext: Boolean = true

      private val tooMany: Int = 1000
    
      override def next(): BigInt =
        n += 1
//      if n > tooMany // this is good for testing
//        throw new IllegalStateException("ok, I'm tired...")
//
        n

  override def contains(n: BigInt): Boolean = n >= 0
